crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2576,7 +2609,53 @@ impl<'db> TypeInferenceBuilder<'db> {
                         Type::Unknown
                     })
             }
-            _ => Type::Todo,
+            (value_ty, slice_ty) => {
+                // Resolve the value to its class.
+                let value_meta_ty = value_ty.to_meta_type(self.db);
+
+                // If the class defines `__getitem__`, return its return type.
+                //
+                // See: https://docs.python.org/3/reference/datamodel.html#class-getitem-versus-getitem
+                let dunder_getitem_method = value_meta_ty.member(self.db, "__getitem__");
+                if !dunder_getitem_method.is_unbound() {
+                    // TODO validate that the `__getitem__` argument adheres to the signature
+                    let CallOutcome::Callable { return_ty } =
+                        dunder_getitem_method.call(self.db, &[slice_ty])
+                    else {
+                        self.dunder_not_callable_diagnostic(
+                            (&**value).into(),
+                            value_ty,
+                            "__getitem__",
```

Comment:
Very hard to test these right now given various TODO types that are returned.

Reply:
Something like this seems to pass for me locally:

```diff
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index 15b2626a8..f74c833ef 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -6829,6 +6829,30 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn subscript_not_callable_getitem() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+                class NotSubscriptable:
+                    __getitem__ = None
+
+                a = NotSubscriptable()[0]
+            ",
+        )?;
+
+        assert_public_ty(&db, "/src/a.py", "a", "Unknown");
+        assert_file_diagnostics(
+            &db,
+            "/src/a.py",
+            &["Dunder method `__getitem__` is not callable on type 'NotSubscriptable'."],
+        );
+
+        Ok(())
+    }
```

Reply:
Yeah, a test like that should work; I'm curious what forms of test you tried, and what Todo types you ran into.

Reply:
I honestly can't remember what I tried but it led to me adding `__call__` support. I think I was looking to resolve a type annotation in a class member and it was returning `Todo`, but I can't remember the details.